js/ui/components/clubhouse.js

```diff
@@ -0,0 +1,1052 @@
+        }
+    }
+
+    setIcon(actor) {
+        actor.add_style_class_name('clubhouse-notification-image');
+        super.setIcon(actor);
+    }
+
+    _rearrangeElements() {
+        let contentBox = this._bodyStack.get_parent();
+        contentBox.add_style_class_name('clubhouse-content-box');
+        let hbox = contentBox.get_parent();
+        let vbox = hbox.get_parent();
+
+        let wrapBin = new St.Bin({ x_fill: true,
+                                   y_fill: true,
+                                   x_align: St.Align.END,
+                                   y_align: St.Align.END,
+                                   x_expand: true,
+                                   y_expand: true });
+        hbox.add_child(wrapBin);
+
+        // A Clutter.BinLayout is used to rearrange the notification
+        // elements in layers:
+        let wrapWidget = new St.Widget({ layout_manager: new Clutter.BinLayout(),
+                                         x_align: Clutter.ActorAlign.FILL,
+                                         y_align: Clutter.ActorAlign.END });
+        wrapBin.set_child(wrapWidget);
+
+        let iconBinParams = { x_fill: false,
+                              y_fill: false,
+                              x_expand: true,
+                              y_expand: true,
+                              x_align: St.Align.END,
+                              y_align: St.Align.END };
+        Object.assign(this._iconBin, iconBinParams);
+
+        let contentBoxParams = { x_expand: true,
+                                 y_expand: false,
+                                 x_align: Clutter.ActorAlign.FILL,
+                                 y_align: Clutter.ActorAlign.END };
+        Object.assign(contentBox, contentBoxParams);
+
+        let actionBinParams = { x_expand: true,
+                                y_expand: true,
+                                x_align: Clutter.ActorAlign.START,
+                                y_align: Clutter.ActorAlign.END };
+        Object.assign(this._actionBin, actionBinParams);
+
+        let expandedLabelActorParams = { y_expand: true,
+                                         y_align: Clutter.ActorAlign.CENTER };
+        Object.assign(this._expandedLabel.actor, expandedLabelActorParams);
+        this._expandedLabel.actor.clutter_text.line_wrap_mode = Pango.WrapMode.WORD_CHAR;
+
+        hbox.remove_child(contentBox);
+        wrapWidget.add_child(contentBox);
+
+        hbox.remove_child(this._iconBin);
+        wrapWidget.add_child(this._iconBin);
+
+        vbox.remove_child(this._actionBin);
+        wrapWidget.add_child(this._actionBin);
+    }
+
+    // Override the callback that changes the button opacity on hover:
+    _sync() {
+        this._closeButton.opacity = 255;
+    }
+
+    _addActions() {
+        // if these attrs are not initialized we're in the constructor
+        if (this._textIdx === undefined || this._textPages === undefined)
+            return;
```

Comment:
How did this work before? (I looked through the diff and it seems that this piece of code is new.)

Reply:
Yes, this is new, but before the rebase we were using Lang for classes and the constructor were called after the `_textIdx` initialization so this didn't crash. With the update to use js classes, the constructor should be called before any reference to `this`, and the parent class call the `_addActions` in the constructor so we need to add this check to avoid a shell crash.

Reply:
I see... In that case I'd prefer if we made `this._inLastPage()` robust to the condition, e.g. make it return `true` when there are no pages/no index.